### ISSUE TITLE
Pytorch -> ONNX -> TensorRT incorrect TensorRT results

### ISSUE THREAD

**Issue by justinblaber** [NVIDIA/TensorRT#813]
I have a trained pytorch model, UNET type architecture that I've already trained. I can convert to ONNX successfully and inference results, although slightly different from pytorch, are very similar. However, when I convert from ONNX to TensorRT, the results are more different and results in an incorrect segmentation.

Pytorch and onnx were both compiled from source to ensure the same cuda and cudnn libraries were used as packaged in tensorrt. The model was, however, trained a few months ago on another machine that had the pip installed version of pytorch.

## Environment

**TensorRT Version**:  TensorRT-7.2.0.14
**GPU Type**: 2070 Max-Q
**Nvidia Driver Version**: 450.51.06
**CUDA Version**: 11.0 (update 1)
**CUDNN Version**:  8.0.2
**Operating System + Version**: ubuntu 18.04 
**Python Version (if applicable)**: 3.7.7
**TensorFlow Version (if applicable)**: 
**PyTorch Version (if applicable)**: 1.7.0
**Baremetal or Container (which commit + image + tag)**: 

## Steps To Reproduce

I can upload a notebook in a bit, but essentially, the following model is loaded:

![model](https://i.imgur.com/TL1IeDo.png)

Inference results with pytorch (correct results):
![pytorch](https://i.imgur.com/xsYY2A5.png)

Model is converted to `onnx` with:

```
orch.onnx.export(model,
                  X.cuda(),
                  'model.onnx',
                  export_params=True,
                  opset_version=11, # Needed for upsample operation
                  verbose=False)
```

Inference results:

![onnx](https://i.imgur.com/EgFzdvs.png)

Note that the topleft pixel value is slightly different from pytorch's:
`5.532728672027588` -> `5.532743453979492`

But overall the segmentation mask is about the same.

Now, for tensort, the model is converted, serialized, then deserialized:

```
import tensorrt as trt
import pycuda.driver as cuda
import pycuda.autoinit

TRT_LOGGER = trt.Logger(trt.Logger.VERBOSE)
model_path = 'model.onnx'

EXPLICIT_BATCH = 1 << (int)(trt.NetworkDefinitionCreationFlag.EXPLICIT_BATCH)
builder = trt.Builder(TRT_LOGGER)
network = builder.create_network(EXPLICIT_BATCH)
parser = trt.OnnxParser(network, TRT_LOGGER)
parser.parse_from_file(model_path)
config = builder.create_builder_config()
config.max_workspace_size = 1 << 20
engine = builder.build_engine(network, config)

engine_path = 'model.engine'
with open(engine_path, 'wb') as f:
    f.write(bytearray(engine.serialize()))

runtime = trt.Runtime(TRT_LOGGER)
with open(engine_path, 'rb') as f:
    engine = runtime.deserialize_cuda_engine(f.read())

h_input = cuda.pagelocked_empty(trt.volume(engine.get_binding_shape(0)), dtype=np.float32)
h_output = cuda.pagelocked_empty(trt.volume(engine.get_binding_shape(1)), dtype=np.float32)
d_input = cuda.mem_alloc(h_input.nbytes)
d_output = cuda.mem_alloc(h_output.nbytes)

stream = cuda.Stream()
context = engine.create_execution_context()
h_input[:] = np.fromfile('X.raw', dtype=np.float32) # Copy preprocessed data to pagelocked memeory

cuda.memcpy_htod_async(d_input, h_input, stream)    # Transfer input data to the GPU.
context.execute_async(bindings=[int(d_input), int(d_output)], stream_handle=stream.handle) # Run inference.
cuda.memcpy_dtoh_async(h_output, d_output, stream)  # Transfer predictions back from the GPU.
stream.synchronize()                                # Synchronize the stream

h_output = h_output.reshape(engine.get_binding_shape(1))
```

I checked terminal output and didnt see any error messages. The inference results I got were:

![tensorrt](https://i.imgur.com/H06wgB9.png)

The inference mask is now wrong (added green blob) and also that topleft value in the mask output is:

```
pytorch  = 5.532728672027588
onnx     = 5.532743453979492
tensortt = 5.454630374908447
```

So now that value is noticeably different, which is concerning because all precision should still be fp32.

**Comment by justinblaber** [comment]
I've experimented around and if I replace the upsampling operation with a transposed convolution it appears to make the differences mostly go away (within fp32 reasonableness)... I actually tested the upsampling operation individually (i.e. model = nn.Upsample), then converted, and the results were the exact same... so that makes me think there's some other deeper rooted issue here.

**Comment by pranavm-nvidia** [comment]
@justinblaber Would you be able to share your model (or a proxy)?

**Comment by ttyio** [comment]
Hello @justinblaber , thanks for reporting

We had some issue for `upsample` when `align_corners = true` and user provide `scale` instead of `size` , more detail I have comment in https://github.com/NVIDIA/TensorRT/issues/823#issuecomment-709896445

I see you did not assign `align_corners` in your pasted `upsample` sample, and the `align_corners = false` by default, we should not hit the bug. However, you did not provide the full model. Could you help check your model to see if all the `upsample` has `align_corners` unassigned? If not you need WAR the issue by code like this `nn.Upsample(size = (14, 20), mode='bilinear', align_corners=True)` before we fix it in TRT.

Thanks.

**Comment by ttyio** [comment]
Close since no response for more than 2 weeks, please reopen if you have more questions, thanks!

**Comment by ttdd11** [comment]
@ttyio I think I am having similar issues here however my model is trained in tensorflow. The results are shifted as above. I think the upsample operation is to blame here. My results are probably worse than this because I have many. Is there a work around for this (#974)?

**Comment by Ekta246** [comment]
I just do not have incorrect results for tensorrt model. I see half of the tensors are 0 while half of them are values.

**Comment by ttyio** [comment]
@Ekta246 , did  you mean the batch 1 is correct and batch 2~N are 0?

**Comment by Ekta246** [comment]
So if the shape of the output array is  (1,54,4,4) then half of the array i.e., (1,27,4,4) are filled with float values while the other half is zeros.

Also, I would just like to add a piece of my Resize node in the ONNX. I read somewhere the Resize should have 4 outputs. But I had managed to write the F.interpolate with size as a argument avoiding the scale.
Could you please tell if something could go wrong `there.
![correct](https://user-images.githubusercontent.com/52435785/106850505-9cf4c280-6682-11eb-90f5-6e7de7626fdb.png)

**Comment by ttyio** [comment]
Hello @Ekta246 , from your screenshot the `pytorch_half_pixel` should work OK with latest 7.2 release. Is the half zeros come from the output of the resize node?

**Comment by Ekta246** [comment]
Ok, I should try TensorRt 7.2.

So in the Onnxruntime if I print the output tensors they are similar to the
output of the Pytorch model. But the TensorRTEngine model has same shape as
that of the Pytorch and onnx model but the values are half zeros and half
correct values.

And by output I meant, it is the final output of the model. Honestly, I
don't know if the problem emerges because of the Resize layer or is it
something else? The good thing is I get the correct output shapes. But why
not the correct values..

Also, could you provide any debugging steps or ideas that I can peek into
if it is the problem with TensorRT Python Api or trtexec. I am still
figuring out.

Sincerely,

**Comment by ttyio** [comment]
Thanks @Ekta246 for trying the latest, the resize output shape always correct, and the error happens in the output values. 

For the debugging tips.
We have another tools https://github.com/NVIDIA/TensorRT/tree/master/tools/Polygraphy, we can use `--trt-outputs mark all`, `--onnx-outputs mark all` to mark all layer as output and compare between onnxruntime and trt.

If your network is large and the layerwise comparison is slow, then you might try manually use https://github.com/NVIDIA/TensorRT/tree/master/tools/onnx-graphsurgeon to modify the onnx network, and still use polygraphy, but this time only compare the output between onnxruntime and TRT, you can do this again and again using binary search to find out the failure node.

**Comment by Ekta246** [comment]
thanks a lot @ttyio, I will have to look into the debugging closely, since I am a novice in this field. 

Also, I haven't worked on TensorRT 7.2. I have worked on 7.1.3.0.
Do you anticipate the 7.2 would do any better not getting the zeros in the arrays.?

Also, do you also mean that irrespective of the correct shape I am getting now, the failure node could be a potential issue in the values when compared to onnx and pytorch model?

Thanks for your huge help and quick response.

**Comment by Ekta246** [comment]
> Thanks @Ekta246 for trying the latest, the resize output shape always correct, and the error happens in the output values.
> 
> For the debugging tips.
> We have another tools https://github.com/NVIDIA/TensorRT/tree/master/tools/Polygraphy, we can use `--trt-outputs mark all`, `--onnx-outputs mark all` to mark all layer as output and compare between onnxruntime and trt.
> 
> If your network is large and the layerwise comparison is slow, then you might try manually use https://github.com/NVIDIA/TensorRT/tree/master/tools/onnx-graphsurgeon to slice the onnx network, and still use polygraphy, but this time only compare the output between onnxruntime and TRT, you can do this again and again using binary search to find out the failure node.

@ttyio Thanks for the crisp explanation. But I cannot understand how do I use the --trt-outputs and --onnx outputs. Are there any example reference samples? I do not understand which files in the https://github.com/NVIDIA/TensorRT/tree/master/tools/Polygraphy
or 
https://github.com/NVIDIA/TensorRT/tree/master/tools/onnx-graphsurgeon
should I refer?

Also, my problem is with after deserializing the .engine file and then running context.execute_v2 I am able to generate output of the model but the values are different than that of the onnxruntime and pytorch model.

**Comment by ttyio** [comment]
Hello @Ekta246 , 
>> Do you anticipate the 7.2 would do any better not getting the zeros in the arrays.?
We keep improving the onnx parser in TRT, but sorry I cannot tell if 7.2 can fix the zeros you see, because it is not root cause yet.

>>  I cannot understand how do I use the --trt-outputs and --onnx outputs. Are there any example reference samples? I do not understand which files in the
The Polygraphy is a tool to compare the TRT result with other framework result. After install you can also get a commandline tool 
E.g, To compare the output between onnxruntime and trt, you can run

    polygraphy run your.onnx --trt --onnxrt 

To mark all nodes in onnx as output and compare between onnxruntime and trt, you can run

    polygraphy run your.onnx --trt --onnxrt --onnx-outputs mark all --trt-outputs mark all

And if you find mark all layer as output is too slow to run, then you can manually edit your onnxfile, you can use onnx-graphsurgeon, or you can use other favorite tools

**Comment by Ekta246** [comment]
I appreciate your answer. Also, just to let you know I am converting a pytorch to onnx and then onnx to a engine through trtexec. 
So I will work out on your steps if I can compare the onnx file and the tensorrt engine file generated through trtexec.

**Comment by Ekta246** [comment]
@ttyio, So I was able to run and I can produce a trace of mismatched nodes. I am still figuring out what could be fixed so as the nodes are not mismatched.

Experiment:
On running fp16 every node was pretty much mismatched
While, for fp32 there were fewer mismatched nodes.

Also, if I can make sense out of it does the mismatched node means the trt does not support that particular onnx node in the .onnx model? Or I am misinterpreting it?
An example of the trace is as below,
Comparing Output: '1268' (dtype=float32, shape=()) with '1268' (dtype=float32, shape=())
[I]         Required tolerances: [atol=0.00039101] OR [rtol=1e-05, atol=0.00037054] OR [rtol=0.00018617, atol=1e-05]
            Runner: trt-runner-N0-02/05/21-01:29:01          | Stats: mean=2.047, min=2.047 at (), max=2.047 at ()
            Runner: onnxrt-runner-N0-02/05/21-01:29:01       | Stats: mean=2.0466, min=2.0466 at (), max=2.0466 at ()
[E]         FAILED | Difference exceeds tolerance (rtol=1e-05, atol=1e-05)

Again, you have been a huge support
TIA

**Comment by huukim911** [comment]
Any find out yet @Ekta246 ? I am getting this kind of mismatch output from trt, too. Rgrds

**Comment by ttanzhiqiang** [comment]
https://github.com/ttanzhiqiang/onnx_tensorrt_project
